kymatio/scattering1d/frontend/base_frontend.py

```diff
@@ -50,31 +62,96 @@ def build(self):
+        self.N_scale = math.ceil(math.log2(self.N))
+
+        # check `pad_mode`, set `pad_fn`
+        if isinstance(self.pad_mode, FunctionType):
+            def pad_fn(x):
+                return self.pad_mode(x, self.pad_left, self.pad_right)
+            self.pad_mode = 'custom'
+        elif self.pad_mode not in ('reflect', 'zero'):
+            raise ValueError(("unsupported `pad_mode` '{}';\nmust be a "
+                              "function, or string, one of: 'zero', 'reflect'."
+                              ).format(str(self.pad_mode)))
+        else:
+            def pad_fn(x):
+                return self.backend.pad(x, self.pad_left, self.pad_right,
+                                        self.pad_mode)
+        self.pad_fn = pad_fn
+
+        # check `normalize`
+        supported = ('l1', 'l2', 'l1-energy', 'l2-energy')
+        if self.normalize not in supported:
+            raise ValueError("unsupported `normalize`; must be one of: %s"
+                             + ", ".join(supported))
+
+        # ensure 2**J <= nextpow2(N)
+        mx = 2**math.ceil(math.log2(self.N))
+        if 2**(self.J) > mx:
+            raise ValueError(("2**J cannot exceed input length (rounded up to "
+                              "pow2) (got {} > {})".format(2**(self.J), mx)))
+
+        # check T or set default
+        if self.T is None:
+            self.T = 2**(self.J)
+        elif self.T == 'global':
+            self.T == mx
+        elif self.T > self.N:
+            raise ValueError(("The temporal support T of the low-pass filter "
+                              "cannot exceed input length (got {} > {})"
+                              ).format(self.T, self.N))
+        self.log2_T = math.floor(math.log2(self.T))
+        self.average_global_phi = bool(self.T == mx)
+        self.average_global = bool(self.average_global_phi and self.average)
 
         # Compute the minimum support to pad (ideally)
-        min_to_pad = compute_minimum_support_to_pad(
-            self.N, self.J, self.Q, r_psi=self.r_psi, sigma0=self.sigma0,
-            alpha=self.alpha, P_max=self.P_max, eps=self.eps,
+        min_to_pad, pad_phi, pad_psi1, pad_psi2 = compute_minimum_support_to_pad(
+            self.N, self.J, self.Q, self.T, r_psi=self.r_psi,
+            sigma0=self.sigma0, alpha=self.alpha, P_max=self.P_max, eps=self.eps,
             criterion_amplitude=self.criterion_amplitude,
-            normalize=self.normalize)
-        # to avoid padding more than N - 1 on the left and on the right,
-        # since otherwise torch sends nans
-        J_max_support = int(np.floor(np.log2(3 * self.N - 2)))
-        self.J_pad = min(int(np.ceil(np.log2(self.N + 2 * min_to_pad))),
-                         J_max_support)
+            normalize=self.normalize, pad_mode=self.pad_mode)
+        if self.average_global:
+            min_to_pad = max(pad_psi1, pad_psi2)  # ignore phi's padding
+
+        J_pad = math.ceil(math.log2(self.N + 2 * min_to_pad))
```

Comment:
can we move the instantiation of `J_pad` out of the frontend? Perhaps it is only necessary to compute it _within_ `scattering_filter_factory` so as to preserve the usability of this function, hiding the padding computation from the user

Reply:
At this point there's lots such computations being done in the frontend. If that kind of refactoring is desired, it can be done separately, but I think keeping some in frontend is reasonable and adds clarity. It'd also require moving steps out of `build()` to after `create_filters()` for `Scattering1D`.

Reply:
then at least let's add a comment saying what `J_pad` does?

Reply:
Why specifically for `J_pad`? It's covered in attribute docstrings along all other attributes

Reply:
i've realized that the `J_pad` logic in the frontend was not enttirely introduced in this commit. 

although it makes it harder to use `scattering_filter_factory` (for demo and visualization purposes), fixing it can be left for another time ...

Reply:
Not a fan of moving compute to a function unless necessary - in this case it's just one line.

> harder to use scattering_filter_factory

How so? It's just `scattering_filter_factory(log2(N), J, Q, T)` - we pass `J_pad` in frontend because that's our "N" for filters. It's also an internal method anyway, and filter length would instead be controlled by `max_pad_factor`

Regardless yes, "optional" refactors best wait

Reply:
> it's just one line

though fair enough if moving all that other stuff, but would need to `return` these